### PR TITLE
Fix #498: Optimize rest arguments

### LINKED ISSUE #498: Optimize rest arguments.
The TypeScript compiler generates the following code for rest arguments:

```
function foo(x, y, z, ...rest) {
  return;
}

function foo(x, y, z) {
  var rest = [];
  for (var i = 0; i < arguments.length - 3; i++) {
    rest[i] = arguments[i + 3];
  }
  return;
}
```

Although `rest` is unused in the body of the function, JS engines can't easily dead code eliminate the allocation of the `rest` array and the subsequent for loop due to the possibility of Array.prototype being mutated:

```
Object.defineProperty(Array.prototype, 42, { set: function() { alert(this[41]); }});
```

It's much easier for the TypeScript compiler to avoid emitting this code when it is not necessary.

See more details:

https://bugzilla.mozilla.org/show_bug.cgi?id=1056446

### PR DESCRIPTION
**Pull request by begincalendar**
Fixes #498.

Inspiration of the implementation is drawn from [this](https://github.com/Microsoft/TypeScript/pull/518#issuecomment-59630642).

### CODE REVIEW

**Comment by begincalendar** [comment]
@mhegazy  I left three failing baselines, because I'm not sure about them.

They are: [parser509668](https://github.com/Microsoft/TypeScript/blob/8a227674ce7853efcf5c6804ded1ed5739883b5c/tests/cases/conformance/parser/ecmascript5/RegressionTests/parser509668.ts), [varArgConstructorMemberParameter](https://github.com/Microsoft/TypeScript/blob/8a227674ce7853efcf5c6804ded1ed5739883b5c/tests/cases/compiler/varArgConstructorMemberParameter.ts) and [restParamModifier2](https://github.com/Microsoft/TypeScript/blob/8a227674ce7853efcf5c6804ded1ed5739883b5c/tests/cases/compiler/restParamModifier2.ts).

Should the rest parameter be emitted when there is a failed attempt to use it as a parameter property?

I'm thinking it's fine for the optimization to occur (i.e. not emit the rest parameter), but I just want to make sure.

File: tests/baselines/reference/accessorWithRestParam.js [hunk 1]
```diff
@@ -9,22 +9,12 @@ var C = /** @class */ (function () {
     function C() {
     }
     Object.defineProperty(C.prototype, "X", {
-        set: function () {
-            var v = [];
-            for (var _i = 0; _i < arguments.length; _i++) {
-                v[_i] = arguments[_i];
-            }
-        },
+        set: function () { },
```
Comment (Kovensky) [inline tests/baselines/reference/accessorWithRestParam.js]: This looks wrong.

This test file is wrong.

`class C { set X(...v) {} }`, which this file is testing, is invalid syntax. `set` methods, on either classes or objects, only accept exactly one valid `FormalParameter`, which does _not_ allow for `FunctionRestParameter`.

Comment (begincalendar) [reply tests/baselines/reference/accessorWithRestParam.js]: Right, thanks.

Yet TypeScript does not allow `set` to have a rest parameter anyway (i.e. the test code won't compile).

So I guess I'll revert this baseline and add it to the list above, where I ask @mhegazy about whether or not to apply the optimization in error cases.

Comment (Kovensky) [reply tests/baselines/reference/accessorWithRestParam.js]: Oh, the transform looks correct -- the test file having a valid-looking output at all is wrong. Invalid syntax should either throw or stay as invalid syntax 😄

**Comment by begincalendar** [comment]
@andy-ms If you get a chance and this applies to you, some input would be much appreciated.

**Comment by andy-ms** [comment]
This can be done without adding a new check flag.
In `shouldAddRestParameter`:
```ts
return !!node
    && !!node.dotDotDotToken
    && node.name.kind === SyntaxKind.Identifier
    && !inConstructorWithSynthesizedSuper
    && resolver.isRestParameterUsed(getOriginalNode(node, isParameter));
```
In `export interface EmitResolver`:
```ts
isRestParameterUsed(node: ts.ParameterDeclaration): boolean;
```
In `checker.ts` `resolveNameHelper`: `if (isUse && result && nameNotFoundMessage && noUnusedIdentifiers && result !== lastNonBlockLocation.symbol) {` --> remove `&& noUnusedIdentifiers`
And in `createResolver`, add `isRestParameterUsed: p => p.symbol!.isReferenced,`.

@rbuckton  Maybe `isReferenced` should be a check flag instead of its own property?

**Comment by andy-ms** [comment]
Is this feature actually worth adding though? What's a real-world use for having a lot of unused rest parameters?

**Comment by begincalendar** [comment]
Thanks for the feedback @andy-ms.
I think your question regarding whether this PR is even necessary is the first priority, but while we wait for clarification on that...

1. Should the unary cast operations in `shouldAddRestParameter` (e.g. `!!node`) be done in another PR?
2. Can we just inline `isRestParameterUsed` (given that node internals are already being accessed within `shouldAddRestParameter` and the interface function would have no other callers)?

**Comment by andy-ms** [comment]
As for whether this is necessary, pinging @rbuckton.

1. I don't think that's necessary because it seems obvious that this was always meant to return a boolean.
2. Sure. I was expecting that it would need to access checker internals, but I guess it doesn't.

**Comment by RyanCavanaugh** [comment]
I think this PR is subsumed by #20307 (fix for #19182)

**Pull request closed without merging**